### ISSUE TITLE
transformItems function has issue with the key name "type"

### ISSUE THREAD

**Issue by wmhopk** [algolia/instantsearch.js#4831]
## 🐛 Bug description

The transformItems function fails silently when using the key _**type**_.  It also appears to prevent any other transforms from happening.

## 🔍 Bug reproduction

**Steps to reproduce the behavior:**

1. Set up the transformItems function as below:
```php
transformItems(items) {
  return items.map(item => ({
    ...item,
    type: item.name.toUpperCase(),
  }));
}
```
or 
```php
transformItems(items) {
  return items.map(item => ({
    ...item,
    type: "Custom Type",
  }));
}
```
2. Create a template using hit.type as a value.
3. In the screenshot below I used the search phrase "District of Columbia Imposes Significant Restrictions on the Hospitality". If you need more results you can use "Background check requirements" as a search term. I've linked a sandbox for testing below.

My workaround was to change the key _**type**_ to _**custom_content_type_key**_ both in the transformItems function and in the result template.

**Live reproduction:**

https://codesandbox.io/s/elated-banach-02mqd?file=/src/app.js

## 💭 Expected behavior

That using  _type_ as a key in tranformItems would either transform the item or fail with an error.

## 🖥 Screenshots

![Screen Shot 2021-08-14 at 2 12 02 PM](https://user-images.githubusercontent.com/282426/129458055-51b7ac67-6422-46a3-b82a-6524600d142a.png)

The screenshot above is with the _type_ parameter replaced with _custom_content_type_key_.

![Screen Shot 2021-08-14 at 2 26 24 PM](https://user-images.githubusercontent.com/282426/129458136-b7715769-beff-46ff-ac6c-1b8d17d16570.png)

The screenshot above uses the _type_ parameter. Notice the News & Analysis label (type) is missing from the left of the title. Also the date is missing to the left of the description.

## Environment

- OS: macOS
- Browser: Chrome
- Version: 92.0.4515.131

## Additional context
**Note:**
I'm using a Drupal module the passes along the content type of a record with the key _**type**_.

**Comment by tkrugg** [comment]
Hi there, thanks for reaching out. 
I've tried your code example and I can't get the behaviour you're describing. https://codesandbox.io/s/zen-yonath-xtouu?file=/src/app.js

I've looked at the codesandbox you provided. I see the issue and for me, it's not quite what you've described. 

It's related to the fact the new value of type is based on its previous value (see my screenshot). 
This is because `transformItems` writes to the state and can receive already transformed items (this is a known problem tracked by https://github.com/algolia/instantsearch.js/issues/4819). 

From what I see on your codesandbox, it's not failing silently, it's just that the templating engine (Hogan) can manage `undefined` values (by showing nothing). 

The workaround you have is the one I would go for in the meantime.

![image](https://user-images.githubusercontent.com/2982512/129706911-6d890981-acd0-4903-ae38-199b5f28ea7a.png)

Closing this as a duplicate of #4819
